proxy/src/pglb/inprocess.rs

```diff
@@ -0,0 +1,465 @@
+            .finish_non_exhaustive()
+    }
+}
+
+impl InProcessUdpSocket {
+    pub fn new_pair() -> (Self, Self) {
+        let (sender_a, receiver_a) = HeapRb::new(16).split();
+        let (sender_b, receiver_b) = HeapRb::new(16).split();
+
+        let addr_a = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 111)), 11111);
+        let addr_b = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 222)), 22222);
+
+        let queue_a = DatagramQueue::new(sender_a, receiver_b);
+        let queue_b = DatagramQueue::new(sender_b, receiver_a);
+
+        let receiver_waker_a = Arc::clone(&queue_a.receiver_waker);
+        let receiver_waker_b = Arc::clone(&queue_b.receiver_waker);
+
+        let socket_a = InProcessUdpSocket {
+            queue: queue_a,
+            sibling_receiver_waker: receiver_waker_b,
+            local_addr: addr_a,
+            remote_addr: addr_b,
+        };
+
+        let socket_b = InProcessUdpSocket {
+            queue: queue_b,
+            sibling_receiver_waker: receiver_waker_a,
+            local_addr: addr_b,
+            remote_addr: addr_a,
+        };
+
+        (socket_a, socket_b)
+    }
+}
+
+impl AsyncUdpSocket for InProcessUdpSocket {
+    #[tracing::instrument(level = "debug", skip_all)]
+    fn create_io_poller(self: Arc<Self>) -> Pin<Box<dyn UdpPoller>> {
+        Box::pin(AlwaysReadyUdpPoller)
+    }
+
+    #[tracing::instrument(level = "debug", skip_all, err)]
+    fn try_send(&self, transmit: &udp::Transmit<'_>) -> io::Result<()> {
+        if transmit.src_ip.is_some_and(|ip| ip != self.local_addr.ip()) {
+            return Err(io::Error::new(
+                io::ErrorKind::AddrNotAvailable,
+                "try_send: address not available",
+            ));
+        }
+        if transmit.destination != self.remote_addr {
+            // TODO: or drop the packet?
+            return Err(io::Error::new(
+                io::ErrorKind::HostUnreachable,
+                "try_send: host unreachable",
+            ));
+        }
+        self.queue.send(transmit.contents)?;
+        Ok(())
+    }
+
+    #[tracing::instrument(level = "debug", skip_all, ret)]
+    fn poll_recv(
+        &self,
+        cx: &mut Context<'_>,
+        bufs: &mut [io::IoSliceMut<'_>],
+        meta: &mut [udp::RecvMeta],
+    ) -> Poll<io::Result<usize>> {
+        if bufs.is_empty() {
+            return Poll::Ready(Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "poll_recv: no buffer",
+            )));
+        }
+        if bufs.len() > 1 {
+            return Poll::Ready(Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "poll_recv: multiple buffers",
+            )));
+        }
+
+        let buf = &mut bufs[0];
+        let len = match self.queue.recv(buf) {
+            Ok(len) => len,
+            Err(e) => {
+                if e.kind() == io::ErrorKind::WouldBlock {
+                    self.sibling_receiver_waker.register(cx);
```

Comment:
Hmm. Could this get into a position that stalls a task?

Task A calls `poll_recv(cx, ...)`, registers its waker.
Task B calls `poll_recv(cx, ...),` replaces the waker.
Task A never wakes up.

I'm not sure if quinn ever gets into that scenario, but there's nothing preventing this situation as far as I can see here.

Reply:
Technically, yes, but quinn uses one receiver task that demuxes the streams, even when we clone the Endpoint. The only way I see this break is if we use the same InProcessUdpSocket to create different Endpoints. That would result in all kinds of breakage.

I changed register() to check that the Waker will wake the same task when there's already a waker registered.